### PR TITLE
Invisible Spray

### PR DESCRIPTION
**Pull request by Shadowmech88**
invisible_spray_display
Invisible spray can be used on any object to render it invisible, or any mob to render that mob's body invisible.
Each can is limited to one use by default.
Invisible spray demo: http://puu.sh/tjOOL/81d725051d.webm
Invisible spray can be removed by application of water, such as with a fire extinguisher.

Invisible objects are no longer visible on a mob's sprite when they are wearing/holding them.

A can of invisible spray can be bought from clown and mime traitor uplinks for 5 TC.

:cl:
 * rscadd: Added invisible spray to clown and mime traitor uplinks for 5 TC. It can be used to render people or objects invisible, and can be removed with water.

### CODE REVIEW

**Comment by Probe1** [comment]
neat but why the clown and not just the mime?

**Comment by Shadowmech88** [comment]
I saw that the super glue was clown+mime and that made me think about the fact that this item is somewhat prank-themed, and so could easily fit with either the clown's or mime's motif.

**Comment by Probe1** [comment]
Kay works for me.

**Comment by ghost** [comment]
>Invisible esword murder
>The spray can't be washed off by showers or something

Why would you do this?

**Comment by Probe1** [comment]
This suddenly got a lot more horrifying.  

While reading this I assumed moving or using it would break its invisibility.  This was a poor assumption.

**Comment by Shadowmech88** [comment]
What if I make it so that any water source washes it off, including spray bottles/extinguishers/splashing?

**Comment by PJB3005** [comment]
... Is this from SpongeBob?

How did they get it off in SpongeBob?

**Comment by Shadowmech88** [comment]
By being splashed with a bucket of water.

**Comment by yclatious** [comment]
While the though of a mime being invisible and doing esword murderboning isn't my taste, the image of sec spraying everything with a extinguisher to find the sneaky mute fucker makes me laugh too much.
Plus, any mime or clown could fill the extinguishers with other shit to make them not work, and sec unwillingly lubing the halls would also be hilarious, promote active antaging and be a great example of sabotage done right.

**Comment by Megamaw** [comment]
honestly, bump up the TC cost a little and make it permanent unless washed off

this will lead to more advanced pranks than what can be pulled off in five minutes

(or maybe an alternate Advanced Invisible Spray)

but either way, A++ i fucking love it

**Comment by Probe1** [comment]
Still too powerful to be able to make weapons or even yourself invisible and go on an invisible murder spree.

**Comment by Aurx** [comment]
Why not make it so that it makes you but NOT what you're wearing invisible?

**Comment by Probe1** [comment]
https://www.youtube.com/watch?v=9OPc7MRm4Y8

File: code/game/objects/items/weapons/cosmetics.dm [hunk 1]
```diff
@@ -190,3 +190,45 @@
...
+		to_chat(user, "\The [src] is empty.")
+		return
+	if(target.invisibility || target.alpha <= 1)
+		to_chat(user, "\The [target] is already invisible!")
+		return
+	if(istype(target, /mob))
+		var/mob/M = target
+		M.alpha = 1	//to cloak immediately instead of on the next Life() tick
+		M.alphas[INVISIBLESPRAY] = 1
+		spawn(invisible_time)
```
Comment (ComicIronic) [inline code/game/objects/items/weapons/cosmetics.dm]: Do not assume the mob still exists after five minutes.

File: code/game/objects/items/weapons/cosmetics.dm [hunk 2]
```diff
@@ -190,3 +190,45 @@
...
+			M.alpha = initial(M.alpha)
+			M.alphas.Remove(INVISIBLESPRAY)
+	else
+		target.alpha = 1
+		spawn(invisible_time)
```
Comment (ComicIronic) [inline code/game/objects/items/weapons/cosmetics.dm]: Do not assume the atom still exists after five minutes.

**Review by ComicIronic — CHANGES_REQUESTED** [review]

File: code/game/objects/items/weapons/cosmetics.dm (hunk 2 above)
Comment (ghost) [reply code/game/objects/items/weapons/cosmetics.dm]: So, just do a check to see if src still exists?

Comment (Shadowmech88) [reply code/game/objects/items/weapons/cosmetics.dm]: To see if `target` still exists.

**Comment by Shadowmech88** [comment]
* When spraying a mob, invisible spray now only makes its *body* invisible, not the items it's wearing or holding. See http://puu.sh/tjOOL/81d725051d.webm for example.
* Invisible spray can now be removed from mobs via application of water, such as from a fire extinguisher.
* Invisible items no longer show up on a mob's body or in a mob's hands.
* Reduced price of invisible spray from 8 TC to 4 TC.

**Comment by Shadowmech88** [comment]
* Invisible items held by a mob no longer show up when examining the mob.

**Comment by murnandreams** [comment]
I think the idea of having it stay until washed off is really good. Maybe keep the telecrystal price at 8 for balance, if that happens. And how does it work with applying it to a backpack, as opposed to putting it in a backpack?

**Comment by Shadowmech88** [comment]
>And how does it work with applying it to a backpack, as opposed to putting it in a backpack?

To put it into a backpack, you must open the backpack's inventory and click an inventory slot. Clicking the backpack itself will spray the backpack.

**Comment by Probe1** [comment]
I still think this is entirely too powerful and should not be merged in its current state.

Acceptable changes in my mind:

Only works on mobs (so clothing, weapons, etc are visible)
Explicitly does not work on weapons
Interacting with anything breaks its invisibility like a cham projector

File: code/modules/mob/living/carbon/human/examine.dm [hunk 3]
```diff
@@ -68,21 +68,21 @@
 	msg += "<EM>[src.name]</EM>!\n"
 
 	//uniform
-	if(w_uniform && !(slot_w_uniform in obscured))
+	if(w_uniform && !(slot_w_uniform in obscured) && !(w_uniform.invisibility || w_uniform.alpha <= 1))
```
Comment (ComicIronic) [inline code/modules/mob/living/carbon/human/examine.dm]: I don't like this copy-paste. Add a proc to do it, like `w_uniform.can_see_equipped()`.

File: code/modules/mob/living/carbon/monkey/update_icons.dm [hunk 4]
```diff
@@ -18,6 +18,19 @@
 
 /mob/living/carbon/monkey/regenerate_icons()
 	..()
+	var/icon/opacity_icon = new(icon, icon_state)
+	if(body_alphas.len)
+		opacity_icon.ChangeOpacity(ARBITRARILY_LARGE_NUMBER)
+		var/lowest_alpha = 255
```
Comment (ComicIronic) [inline code/modules/mob/living/carbon/monkey/update_icons.dm]: Don't particularly like how often this loop is copypasted. Make it a proc like `get_lowest_alpha()`.

**Review by ComicIronic — CHANGES_REQUESTED** [review]

**Comment by ComicIronic** [comment]
It's almost there.

**Review by ComicIronic — APPROVED** [review]

**Comment by StreptoMutans** [comment]
What happens if you apply it to a shard?

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 5]
```diff
@@ -330,6 +334,12 @@ var/global/list/damage_icon_parts = list()
 		else
 			warning("Invalid h_style for [species.name]: [h_style]")
 
+	if(body_alphas.len)
```
Comment (ComicIronic) [inline code/modules/mob/living/carbon/human/update_icons.dm:337]: Shouldn't this still use `get_lowest_body_alpha()`?

File: code/modules/mob/living/carbon/monkey/examine.dm [hunk 6]
```diff
@@ -2,18 +2,19 @@
 
 	var/msg = "<span class='info'>*---------*\nThis is [bicon(src)] \a <EM>[src]</EM>!\n"
 
-	if (src.handcuffed)
+	if (src.handcuffed && !(handcuffed.invisibility || handcuffed.alpha <= 1))
```
Comment (ComicIronic) [inline code/modules/mob/living/carbon/monkey/examine.dm]: Forgot to use `is_visible()` in this file.

File: code/modules/mob/living/carbon/human/update_icons.dm (hunk 5 above)
Comment (Shadowmech88) [reply code/modules/mob/living/carbon/human/update_icons.dm:337]: It should, I forgot about this one.

**Review by ComicIronic — DISMISSED** [review]
Just need to fix your proc use.

**Comment by Shadowmech88** [comment]
A supermatter shard? It works as normal.

File: code/modules/mob/living/carbon/human/update_icons.dm (hunk 5 above)
Comment (Shadowmech88) [reply code/modules/mob/living/carbon/human/update_icons.dm:337]: Done.

**Comment by Megamaw** [comment]
i'd like to reiterate my stance: making it last only five minutes is a good way to make it only useful for a quick killing spree, rather than actual pranks

i mean, if there's still attack messages from invisible people, we can just arm up security with cleaning equipment

tl;dr i think making last as indefinitely until washed a: keeps to source material better and b: allows applications other than stealth murderbone

**Comment by Shadowmech88** [comment]
Making it indefinite would allow someone to use it to hide essentially forever.

**Comment by Megamaw** [comment]
unless they got doused, and if they stayed quiet. iirc making noise will still blow your cover-- so opening doors to move to other parts of the hall, or interacting with closets or anything really

heck, your backpack makes noise.

if you're able to circumvent all these things and evade a bunch of fucks out for you with space cleaner, you deserve to be rewarded.

**Comment by murnandreams** [comment]
I agree with mega, the prospect of security going on a search, needing to work to get cleaner and such will add a whole new demension of gameplay. Increase the cost to balance it. It'd be interesting for all the reasons mega pointed out earlier.

**Comment by Megamaw** [comment]
also mind you this is a mime/clown item only. it's not like literally every traitor can buy this.

**Comment by murnandreams** [comment]
You have to be completely naked as well, unless you purchase more.

**Comment by Shadowmech88** [comment]
>unless they got doused, and if they stayed quiet

This is the issue, that they'd have to make themselves noticed in order to be found. If someone is visible, you can have the main station keep an eye out and Security search Maintenance bit by bit, and you're decently likely to find them.
If they're invisible forever and don't want to be found, what hope would you have? If there's more than just you in the room, something like a backpack ruffling wouldn't be a tip off that there's an invisible man there. The AI isn't even capable of tracking them. You could probably lay down in the corner of the bar and never be found, much less if you were actively avoiding Security.

These things can be varedited, feel free to ask an admin to take all your TC in exchange for making the `invisible_time` var ludicrously large.

**Comment by murnandreams** [comment]
So the point is, if they do nothing as you suggest, they aren't antagging and thus not attracting security.

**Comment by Shadowmech88** [comment]
They could also do stuff that doesn't make it obvious there's an invisible person doing them. They could spray themselves and a backpack and tailgate anyone into anywhere and, once they're alone, take whatever they want and flush themselves down disposals.
At least this way the station will only have to endure that for a short time, if someone decides to do so.

**Comment by Megamaw** [comment]
so let me get this straight, mr. 88

it's broken if it has any practical application for the traitor?

five minutes isn't a lot of time to do anything interesting other than murderbone.

**Comment by PJB3005** [comment]
I think infinite would be better yes.

Honestly this is the kind of thing we can't predict and the best we can do is to push it in and rebalance it if it's proving to be too OP.

I say make it infinite and ship it.

**Comment by Shadowmech88** [comment]
How much should it cost, then?

**Comment by murnandreams** [comment]
5 TC, max of four invisible items

**Comment by Shadowmech88** [comment]
Alright, we'll see.

**Comment by Shadowmech88** [comment]
* The invisible spray that can be purchased from mime and clown traitor uplinks is now an infinite variant that will never wear off on its own.

**Comment by Intigracy** [comment]
Water removes it, right?

**Comment by Shadowmech88** [comment]
From mobs, yes.

It doesn't remove it from items because I couldn't think of a good way to track whether an item has been sprayed with invisible spray, since items don't have a list of alphas that gets regularly checked like mobs do in their Life().

**Comment by Probe1** [comment]
That is even worse than before.

**Comment by Lokiamis** [comment]
Anybody who thinks it should be infinite doesn't remember the cloaking device and when we had to have thermals in security. 
-requires heavy metagaming to counter. 
-invisible bodies are essentially removed from the game even if they're in the middle of the bar
-free escape if objective is completed with <15 tc
-at least prevent it from being used on the disk
-ability to buy 4 covers jumpsuit(pockets), weapon, backpack, radio. 
-can spray your objective items, spare ID, or whatever and security will never know even if they catch you
I'm not crazy and TCs/costs were doubled, right? 5tc now is equal to 2.5tc befote, making this one of the cheapest syndicate items.

**Comment by Exxion** [comment]
The cost is a good point
Personally, I'd say make it 10 or more

As for the rest, I play neither antag nor jobs from whom antags commonly steal, so I'm not exactly an authority

**Comment by murnandreams** [comment]
I think it's important to find a way for water to wash away the spray from items. That will be a huge issue if there's no way to counter against items. How do monkey cubes change from an object to a mob, with water? Maybe something similar can be done to make washing it off items possible.

**Comment by Shadowmech88** [comment]
>How do monkey cubes change from an object to a mob, with water?

Monkey cubes specifically have an interaction with water.
The issue isn't making code execute when an item contacts water, it's keeping track of whether an item has been sprayed with invisible spray. I guess I could make an `/obj`-level var for it, but I don't like the idea of making a new var for just this one specific interaction with this one specific item.

**Comment by Shadowmech88** [comment]
* Invisible spray can now be washed off of objects with water.

**Comment by N3X15** [comment]
Conflicts, must be fixed before merge.

**Comment by Shadowmech88** [comment]
Conflicts fixed.
